File: addingwords/addingwords.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.HashMap;

/**
 * addingwords
 */
public class addingwords {
    
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        HashMap<String,Integer> charsToNums = new HashMap<>();
        HashMap<Integer,String> numsToChars = new HashMap<>();

        while (true) {
            String line = br.readLine();
            String[] exp;
            if(line == null) break;
            else exp = line.split(" ");

            switch (exp[0]) {
                case "def":
                    numsToChars.remove(charsToNums.get(exp[1])); //Essential: Remove the old value associated with the variable from the reverse hashmap!
                    charsToNums.put(exp[1], Integer.parseInt(exp[2]));
                    numsToChars.put(Integer.parseInt(exp[2]), exp[1]);
                    break;
                case "calc":
                    String[] calculation = Arrays.copyOfRange(exp, 1,exp.length);
                    Integer result = charsToNums.get(calculation[0]);
                    if(result == null){
                        System.out.println(String.join(" ",calculation)+" unknown");
                        break;
                    }
                    for (int i = 2; i < calculation.length; i+=2) {
                        Integer number = charsToNums.get(calculation[i]);
                        if(number == null) {
                            result = 1000000000;
                            break;
                        }
                        if(calculation[i-1].equals("+")) result += number;
                        if(calculation[i-1].equals("-")) result -= number;
                    }
                    System.out.println(String.join(" ",calculation) + (numsToChars.get(result) == null ? " unknown" : " " + numsToChars.get(result)));
                    break;
                case "clear":
                    charsToNums = new HashMap<>();
                    numsToChars = new HashMap<>();
                    break;
                default:
                    break;
            }
        }
    }
    
}

File: addingwords/addingwords.py
from sys import stdin
var_to_val = dict()
val_to_var = dict()


for line in stdin:
    line_list = line.strip().split()
    finit = False
    if line_list[0] == "def":
        if line_list[1] in var_to_val:
            del val_to_var[var_to_val[line_list[1]]]
        var_to_val[line_list[1]] = line_list[2]
        val_to_var[line_list[2]] = line_list[1]
    elif line_list[0] == "calc":
        final_exp = []
        for elem in line_list[1:]:
            if elem == "=":
                break
            if elem == '+' or elem == '-':
                final_exp.append(elem)
            else:
                value = var_to_val.get(elem)
                if value is None:
                    print(" ".join(line_list[1:]) + " unknown")
                    finit = True
                    break
                else:
                    final_exp.append(value)
        if not finit:
            result = str(eval("".join(final_exp)))
            if val_to_var.get(result) is None:
                print(" ".join(line_list[1:]) + " unknown")
            else: print(" ".join(line_list[1:]) + " " + val_to_var.get(result))
    elif line_list[0] == "clear":
        var_to_val = dict()
        val_to_var = dict()

